### PR TITLE
spec: add AliasReassignment

### PR DESCRIPTION
**Pull request by WalterBright**
Implementation: https://github.com/dlang/dmd/pull/11846/

### CODE REVIEW

File: spec/declaration.dd
Comment (ibuclaw) [inline spec/declaration.dd:540]: Citation? I was under the impression that this is currently inconclusive.

**Comment by 12345swordy** [comment]
@atilaneves there is a lot of spec pull requests that @WalterBright  has made that needs review from you as you took over @andralex leadership position on this. Can you spare some time and review this PR and other PR likes this that Walter has made?

File: spec/declaration.dd
Comment (atilaneves) [reply spec/declaration.dd:540]: So was I.

Comment (WalterBright) [reply spec/declaration.dd:540]: I don't need a test to know that:
```
malloc(5);
```
is faster and uses less memory than:
```
malloc(5);
malloc(5);
```
:-)

Comment (maxhaton) [reply spec/declaration.dd:540]: https://github.com/dlang/phobos/pull/7756#issuecomment-764048209 Some plots on this. The performance increase is measurable but not particularly dramatic for `staticMap`

**Comment by mdparker** [comment]
@atilaneves The text looks good editorially. I'm good for merging this if you and @ibuclaw are good now with the objection you raised.

**Comment by Imperatorn** [comment]
Merge?

File: spec/declaration.dd
Comment (BorisCarvajal) [reply spec/declaration.dd:540]: > I don't need a test to know that:
> 
> ```
> malloc(5);
> ```
> 
> is faster and uses less memory than:
> 
> ```
> malloc(5);
> malloc(5);
> ```
> 
> :-)


But that's exactly what `static foreach` is doing, it will `syntaxCopy` the `foreach` body N-times.
The rationale should simply be to provide an imperative alternative for constructing alias sequences vs the recursive way.
The performance is just an implementation detail. I hope it gives more performance though.

**Comment by maxhaton** [comment]
> Merge?

Needs WalTila approval. 

@WalterBright @atilaneves 

It's been something like 7 months now, needs a decision. 

There are probably lessons to be learned from how this has been handled.

I say just merge, if something bad happens due to Alias Assign down the line it will likely be due to the implementation, and if we decide it's bad language design then the design itself can be deprecated reasonably smoothly as its use should be both fairly limited and mostly by the hands of people who know what they're doing and are in either maintainers or in direct contact with maintainers.

**Comment by atilaneves** [comment]
Does anyone have numbers on the actual impact on compile-times due to this?

**Comment by mdparker** [comment]
@maxhaton So what's the verdict? Is the performance claim good to go?

**Comment by andralex** [comment]
FWIW in all my real-world testing (building large apps) AliasAssignment has been a win or break even, with huge gains in code size and readability.

**Comment by mdparker** [comment]
Then I propose:

```
$(RATIONALE Alias reassignment can result in faster compile times and lowered memory consumption, and requires significantly fewer lines of code than the alternative recursive method.)
```

**Comment by andralex** [comment]
> Then I propose:
> 
> ```
> $(RATIONALE Alias reassignment can result in faster compile times and lowered memory consumption, 
> and requires significantly fewer lines of code than the alternative recursive method.)
> ```

s/fewer lines of/simpler/

**Comment by mdparker** [comment]
Paging @WalterBright.

**Comment by mdparker** [comment]
> s/fewer lines of/simpler/

Then perhaps also s/requires/results in/.

**Comment by maxhaton** [comment]
> FWIW in all my real-world testing (building large apps) AliasAssignment has been a win or break even, with huge gains in code size and readability.

I agree/concur with Andrei. This feature is still a little raw (Different syntax? It's a bit weird as is in the sense that it's not quite an assignment given the restrictions), but the code-size wins are really nice, and the blow-up losses seem to be an implementation problem which we can nail down (they are _awful_ when you hit them, but we're talking staticMaps over thousands and thousands of elements here).

**Comment by atilaneves** [comment]
> FWIW in all my real-world testing (building large apps) AliasAssignment has been a win or break even, with huge gains in code size and readability.

How big was the median win?

**Comment by andralex** [comment]
> > FWIW in all my real-world testing (building large apps) AliasAssignment has been a win or break even, with huge gains in code size and readability.
> 
> How big was the median win?

Say some 5%-7% but on a rather noisy setup that measured only the whole project bottom line. Let me know if precision is needed and I can set up a more rigorous experimental rig.

**Comment by CyberShadow** [comment]
Either the compiler behavior needs to be put behind a `-preview` flag (I'm guessing not doable because Phobos is using this) or this needs to be merged, because right now the reference implementation's behavior does not correspond to the official specification.

We can help prevent incidents like these from happening in the future by requiring (i.e. enforcing as part of CI) that the compiler test suite / Phobos / etc. are valid according to the official grammar. This will require first making a pull request to dlang.org, describing the language change, which only then will allow any changes to the reference implementation (DMD) to be merged.

File: spec/declaration.dd
Comment (mdparker) [reply spec/declaration.dd:540]: @WalterBright The consensus appears to be to revise this bit as:

> $(RATIONALE Alias reassignment can result in faster compile times and lowered memory consumption, 
and requires significantly simpler code than the alternative recursive method.)

**Review by mdparker — DISMISSED** [review]

**Review by mdparker — APPROVED** [review]

File: spec/declaration.dd
Comment (CyberShadow) [inline spec/declaration.dd:512]: Sorry, how is this different from the `AliasAssign` rule added in #2868?

Furthermore, `AliasReassignment` is not referenced anywhere in the spec, which effectively makes this syntax not a part of the language. What does this grammar block describe exactly?

@WalterBright

File: spec/declaration.dd
Comment (CyberShadow) [inline spec/declaration.dd:510]: `Type` is declared in `type.dd`, not in this file.

File: spec/declaration.dd
Comment (CyberShadow) [inline spec/declaration.dd:512]: Ditto for `BasicType`.

Comment (CyberShadow) [reply spec/declaration.dd:512]: @WalterBright ping

Comment (CyberShadow) [reply spec/declaration.dd:512]: @WalterBright ping

**Pull request merged**
